File: src/test_kdl.cpp
#include <ros/ros.h>
#include <urdf/model.h>
#include <kdl/chainjnttojacsolver.hpp>
#include <kdl/kdl.hpp>
#include <kdl_parser/kdl_parser.hpp>
#include <random>

int main(int argc, char** argv)
{
  ros::init(argc, argv, "test_kdl_cpp");
  ros::NodeHandle nh("~");

  std::string base_link, eef;
  if (!nh.getParam("base_link", base_link))
  {
    ROS_ERROR("Missing base_link");
    return -1;
  }

  if (!nh.getParam("eef", eef))
  {
    ROS_ERROR("Missing eef");
    return -1;
  }

  int max_trials;
  if (!nh.getParam("max_trials", max_trials))
  {
    ROS_ERROR("Missing max_trials");
    return -1;
  }

  if (max_trials <= 0)
  {
    ROS_ERROR("max_trials <= 0");
    return -1;
  }

  urdf::Model model;
  if (!model.initParam("/robot_description"))
  {
    ROS_ERROR("Missing /robot_description");
    return -1;
  }
  KDL::Tree tree;
  KDL::Chain chain;
  kdl_parser::treeFromUrdfModel(model, tree);
  tree.getChain(base_link, eef, chain);
  KDL::ChainJntToJacSolver jac_solver(chain);
  KDL::JntArray q(chain.getNrOfJoints());
  std::uniform_real_distribution<double> unif(-M_PI, M_PI);
  std::default_random_engine re;
  KDL::Jacobian out(chain.getNrOfJoints());
  ROS_INFO("Initialized test_kdl_cpp!");
  ros::Time init = ros::Time::now();

  for (unsigned int t = 0; t < max_trials; t++)
  {
    for (unsigned int i = 0; i < chain.getNrOfJoints(); i++)
    {
      q(i) = unif(re);
    }

    jac_solver.JntToJac(q, out);
  }

  ROS_INFO_STREAM("Elapsed time: " << (ros::Time::now() - init).toSec());

  return 0;
}

File: scripts/test_kdl.py
#!/usr/bin/env python
import rospy
import math
import random
import PyKDL as kdl
from kdl_parser_py import urdf


if __name__ == "__main__":
    rospy.init_node("test_kdl_py")

    if not rospy.has_param("~base_link"):
        rospy.logerr("Missing base_link")
        exit()

    if not rospy.has_param("~eef"):
        rospy.logerr("Missing eef")
        exit()

    if not rospy.has_param("~max_trials"):
        rospy.logerr("Missing max_trials")
        exit()

    base_link = rospy.get_param("~base_link")
    eef = rospy.get_param("~eef")
    max_trials = rospy.get_param("~max_trials")

    tree = urdf.treeFromParam("/robot_description")

    if not tree[0]:
        rospy.logerr("Missing /robot_description")
        exit()

    chain = tree[1].getChain(base_link, eef)
    jac_solver = kdl.ChainJntToJacSolver(chain)
    q = kdl.JntArray(chain.getNrOfJoints())
    J = kdl.Jacobian(chain.getNrOfJoints())

    rospy.loginfo("Initialized test_kdl_py")
    init_time = rospy.Time.now()
    for i in range(max_trials):
        for j in range(chain.getNrOfJoints()):
            q[j] = random.uniform(-math.pi, math.pi)

        jac_solver.JntToJac(q, J)

    rospy.loginfo("Elapsed time: " + str((rospy.Time.now() - init_time).to_sec()))
